File: 백준/Gold/2493. 탑/탑.py
import sys

N = int(input())

Nlist = list(map(int,sys.stdin.readline().split()))


stack = []
answer = []

for i in range(N) :

    while stack:

        if stack[-1][1] <= Nlist[i] :
            stack.pop()

        else :
            #찾았다
            answer.append(stack[-1][0])
            break
    if len(stack) == 0 :
        answer.append(0)

    stack.append([i+1,Nlist[i]])

for i in range(N):
    print(answer[i], end= " ")
    

File: 백준/Gold/2493. 탑/탑.java
import java.util.*;
import java.lang.*;
import java.io.*;

// The main method must be in a class named "Main".
class Main {
    static int N = 0;
    static int answer = 0;
    
    public static void main(String[] args) throws IOException{
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        BufferedWriter bw = new BufferedWriter(new OutputStreamWriter(System.out));

        StringBuilder sb = new StringBuilder();
        N= Integer.parseInt(br.readLine());
        String [] sarr = br.readLine().split(" ");
        int [] Narr = new int [N];
        for(int i= 0 ; i<N; i++)
            {
                Narr[i] = Integer.parseInt(sarr[i]);
            }
        int [] answerarr = new int[N];
        Stack<int [] > stack = new Stack<>();

        for (int x = 0 ; x<N; x++)
            {
                int idx = x;
                int val = Narr[x];

                while(!stack.isEmpty() && stack.peek()[1] < val)
                    {
                        stack.pop();
                    }

                if(stack.isEmpty())
                {
                    answerarr[idx] = 0;
                }
                else
                {
                    answerarr[idx] = stack.peek()[0]+1;
                }
                stack.push(new int [] {idx,val});
            }

        for (int i = 0 ; i<N; i++)
            {
                sb.append(answerarr[i]+" ");
            }

        bw.write(sb.toString());
        bw.close();
    }
}